### DISCUSSION TITLE
Setup Zitadel as kubectl oidc provider

### DISCUSSION THREAD

**Discussion by draeron** [zitadel/zitadel#7042] [General]
I'm looking / experimenting using zitadel as a oidc provider for kubeapi / kubectl and I can't really get is working properly. My understanding of why it doesn't works is that you can't get a claim with a list of role / group.

So this the scope `urn:zitadel:iam:org:project:roles` you get something like this

```json
  "urn:zitadel:iam:org:project:1111111111:roles": {
    "key": {
      "11111111111111": "zitadel.local"
    }
  },
```

But kubernetes api doc says `--oidc-groups-claim` arguments says this: 

> JWT claim to use as the user's group. If the claim is present it must be an array of strings.

Anyone had any success ? Really surprised this can't work out of the box since there are documentation on running zitadel inside a k8s cluster...

Ref:
- [Configuring the API Server](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#configuring-the-api-server)
- [oidc-login](https://github.com/int128/kubelogin/blob/master/docs/setup.md#5-set-up-the-kubeconfig)

**Comment by hifabienne** [comment]
Hei @draeron 
You can configure a ZITADEL Action to put the roles into the token the way you need it.
We do have an example of an action here: https://github.com/zitadel/actions/blob/main/examples/custom_roles.js

**Comment by Smana** [comment]
I'm also interested in configuring groups based on roles. I managed to give permissions based on the email, but now I'm trying to use roles. Still digging...

**Comment by draeron** [comment]
this is what I ended as a action script.

````js
/**
 * sets the roles an additional claim in the token with roles as value an project as key
 * 
 * The role claims of the token look like the following:
 * 
 * // added by the code below
 * "my:zitadel:grants": ["{projectId}:{roleName}", "{projectId}:{roleName}", ...],
 * // added automatically
 * "urn:zitadel:iam:org:project:roles": {
 *   "asdf": {
 *     "201982826478953724": "zitadel.localhost"
 *   }
 * }
 *
 * Flow: Complement token, Triggers: Pre Userinfo creation, Pre access token creation
 *
 * @param ctx
 * @param api
 */
let logger = require("zitadel/log")
function flatRoles(ctx, api) {
  logger.log("user: " + JSON.stringify(ctx.v1.user))
  
  if (ctx.v1.user.grants == undefined || ctx.v1.user.grants.count == 0) {
    return;
  }

  let grants = [];
  ctx.v1.user.grants.grants.forEach(claim => {
    claim.roles.forEach(role => {
        grants.push(role)  
    })
  })

  api.v1.claims.setClaim('roles', grants)
  api.v1.claims.setClaim('groups', grants)
}```

**Reply by Pyrrha** [reply]
I've found a similar script (almost the same than yours) on [ArgoCD configuration](https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/zitadel/#setting-up-an-action-in-zitadel).
Following those steps with trigger only at user info, not token, should do the trick.

For current uses-cases, you can only choose groups, which is default for some platforms. Or maybe you've a specific usage with roles?

Here's my [RBAC related policy](https://github.com/Pyrrha/home-cluster/blob/68d60eeecf294d94cc5c51d6596b2ba1ae0c02cb/component-config/rbac/rbac.yaml)
Api-server is configured with `--oidc-groups-claim=groups`.

**Comment by Smana** [comment]
Great! that works fine, thank you @draeron 

```console
echo "eyJ..." | jwt decode -

...
  "groups": [
    "admin",
    "manager"
  ],
  "iat": 1731751985,
  "iss": "https://<REDACTED>",
  "nonce": "...",
  "roles": [
    "admin",
    "manager"
  ],
...
}
```

For those interested, [here](https://github.com/Smana/cloud-native-ref/tree/main/security/base/zitadel)'s how I deploy Zitadel (Crossplane composition, CloudNativePG, Cert-manager, Gateway-API...).

I guess this issue could be closed?

**Reply by Smana** [reply]
I'll write a blog post on Zitadel and how I'm using it with Kubernetes/Headlamp... . I just want to figure out if that'd be possible to fetch Google groups and map to roles if that's possible but I'll open another issue.
